File: 2n/algorismica/P2-Batalla-a-l'illa/batalla.py
""""Resolució al problema de la batall a l'illa"""
import sys
import math
from functools import wraps

import time

def time_cost(func):
    """Decorador per mesurar el temps d'execució d'una funció."""
    @wraps(func)
    def wrapper(*args, **kwargs):
        start = time.perf_counter()
        result = func(*args, **kwargs)
        end = time.perf_counter()
        print(f"Execution time of {func.__name__}: {end - start:.6f} seconds")
        return result
    return wrapper


def llegir_entrada(nom_fitxer):
    """Llegeix l'entrada del fitxer i retorna el nombre de nois i les seves posicions."""
    with open(nom_fitxer, 'r', encoding='utf-8') as f:
        n = int(f.readline())
        punts = [tuple(map(int, línia.split())) for línia in f]
    return n, punts

def escriure_sortida(nom_fitxer, distancia, equip1):
    """Escriu la sortida al fitxer amb la distància màxima i els jugadors de l'equip roig."""
    with open(nom_fitxer, 'w', encoding='utf-8') as f:
        f.write(f"{distancia:.6f}\n")
        f.write(' '.join(map(str, sorted(equip1))) + '\n')

@time_cost
def resoldre(n, punts):
    """Resoldre el problema de la batalla de nois per maximitzar la distància entre els equips."""
    total = 2 * n
    millor = [0, 0]  # [distància, màscara_roig]

    # Generador que produeix els índexs d'un equip a partir d'una màscara de bits
    def indices_equip(mask):
        for i in range(total):
            if (mask >> i) & 1:
                yield i
    
    # Generador que produeix totes les parelles de distàncies entre equips
    def distancies_entre_equips(roig_mask, blau_mask):
        for i in indices_equip(roig_mask):
            for j in indices_equip(blau_mask):
                yield math.hypot(punts[i][0] - punts[j][0], punts[i][1] - punts[j][1])
    
    def min_dist(roig_mask, blau_mask):
        # Trobem la distància mínima usant el generador
        min_d = float('inf')
        for d in distancies_entre_equips(roig_mask, blau_mask):
            if d < min_d:
                min_d = d
                if min_d <= millor[0]:
                    return 0  # sortida anticipada
        return min_d

    def count_bits(x):
        return bin(x).count("1")

    def back(i, roig_mask, blau_mask):
        if count_bits(roig_mask) > n or count_bits(blau_mask) > n:
            return
        if i == total:
            if count_bits(roig_mask) == n:
                d = min_dist(roig_mask, blau_mask)
                if d > millor[0]:
                    millor[0] = d
                    millor[1] = roig_mask
            return

        # Assigna el jugador i a l'equip roig
        back(i + 1, roig_mask | (1 << i), blau_mask)
        # Assigna el jugador i a l'equip blau
        back(i + 1, roig_mask, blau_mask | (1 << i))

    # Trencament de simetria: fixa el jugador 0 a l'equip roig
    back(1, 1 << 0, 0)

    # Descodifica la màscara de bits en una llista
    equip_roig = [i + 1 for i in range(total) if (millor[1] >> i) & 1]
    return millor[0], equip_roig



def main():
    """Funció principal per llegir l'entrada, resoldre el problema i escriure la sortida."""
    if len(sys.argv) >= 3:
        fitxer_entrada = sys.argv[1]
        fitxer_sortida = sys.argv[2]
    else:
        fitxer_entrada = sys.stdin
        fitxer_sortida = sys.stdout

    n, punts = llegir_entrada(fitxer_entrada)
    distancia, equip1 = resoldre(n, punts)
    escriure_sortida(fitxer_sortida, distancia, equip1)

if __name__ == "__main__":
    main()

File: 2n/algorismica/P2-Batalla-a-l'illa/batalla.hs
-- Resolució al problema de la batalla a l'illa
import System.Environment (getArgs)
import System.IO
import Data.List (sort)
import Data.Bits
import Text.Printf
import System.CPUTime

-- Tipus de dades
type Punt = (Int, Int)
type Mascara = Int

-- Decorador per mesurar temps d'execució
timeCost :: String -> IO a -> IO a
timeCost name action = do
    start <- getCPUTime
    result <- action
    end <- getCPUTime
    let diff = fromIntegral (end - start) / (10^12)
    printf "Execution time of %s: %.6f seconds\n" name (diff :: Double)
    return result

-- Llegeix l'entrada del fitxer
llegirEntrada :: String -> IO (Int, [Punt])
llegirEntrada nomFitxer = do
    contents <- readFile nomFitxer
    let linies = lines contents
        n = read (head linies)
        punts = map ((\[x,y] -> (x,y)) . map read . words) (tail linies)
    return (n, punts)

-- Escriu la sortida al fitxer
escriureSortida :: String -> Double -> [Int] -> IO ()
escriureSortida nomFitxer distancia equip1 = do
    writeFile nomFitxer $ unlines 
        [ printf "%.6f" distancia
        , unwords (map show (sort equip1))
        ]

-- Calcula la distància euclidiana entre dos punts
distancia :: Punt -> Punt -> Double
distancia (x1, y1) (x2, y2) = sqrt $ fromIntegral ((x1-x2)^2 + (y1-y2)^2)

-- Genera els índexs d'un equip a partir d'una màscara de bits
indicesEquip :: Mascara -> Int -> [Int]
indicesEquip mask total = [i | i <- [0..total-1], testBit mask i]

-- Calcula totes les distàncies entre dos equips
distanciesEntreEquips :: [Punt] -> Mascara -> Mascara -> Int -> [Double]
distanciesEntreEquips punts roigMask blauMask total = 
    [ distancia (punts !! i) (punts !! j) 
    | i <- indicesEquip roigMask total
    , j <- indicesEquip blauMask total
    ]

-- Troba la distància mínima entre dos equips
minDist :: [Punt] -> Mascara -> Mascara -> Int -> Double -> Double
minDist punts roigMask blauMask total millor = 
    case distanciesEntreEquips punts roigMask blauMask total of
        [] -> 0
        dists -> foldl1 min dists

-- Compta els bits activats en una màscara
countBits :: Mascara -> Int
countBits = popCount

-- Algoritme de backtracking per trobar la millor assignació
backtrack :: [Punt] -> Int -> Int -> Int -> Mascara -> Mascara -> (Double, Mascara) -> (Double, Mascara)
backtrack punts n total i roigMask blauMask millor@(millorDist, _)
    | countBits roigMask > n || countBits blauMask > n = millor
    | i == total = 
        if countBits roigMask == n
        then let d = minDist punts roigMask blauMask total millorDist
             in if d > millorDist then (d, roigMask) else millor
        else millor
    | otherwise = 
        let -- Assigna el jugador i a l'equip roig
            millor1 = backtrack punts n total (i+1) (setBit roigMask i) blauMask millor
            -- Assigna el jugador i a l'equip blau
            millor2 = backtrack punts n total (i+1) roigMask (setBit blauMask i) millor1
        in millor2

-- Resol el problema principal
resoldre :: Int -> [Punt] -> IO (Double, [Int])
resoldre n punts = do
    let total = 2 * n
        -- Trencament de simetria: fixa el jugador 0 a l'equip roig
        (millorDist, millorMask) = backtrack punts n total 1 (setBit 0 0) 0 (0, 0)
        equipRoig = [i + 1 | i <- [0..total-1], testBit millorMask i]
    return (millorDist, equipRoig)

-- Funció principal
main :: IO ()
main = do
    args <- getArgs
    (fitxerEntrada, fitxerSortida) <- case args of
        [entrada, sortida] -> return (entrada, sortida)
        _ -> do
            putStrLn "Ús: programa fitxer_entrada fitxer_sortida"
            return ("entrada.txt", "sortida.txt")
    
    (n, punts) <- llegirEntrada fitxerEntrada
    (distanciaResult, equip1) <- timeCost "resoldre" (resoldre n punts)
    escriureSortida fitxerSortida distanciaResult equip1